pkg/backend/filestate/lock.go

```diff
@@ -0,0 +1,350 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filestate
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/pulumi/pulumi/pkg/apitype"
+	"github.com/pulumi/pulumi/pkg/backend"
+	"github.com/pulumi/pulumi/pkg/engine"
+	"github.com/pulumi/pulumi/pkg/operations"
+	"github.com/pulumi/pulumi/pkg/resource/config"
+	"github.com/pulumi/pulumi/pkg/tokens"
+	"github.com/pulumi/pulumi/pkg/util/contract"
+	"github.com/pulumi/pulumi/pkg/util/fsutil"
+	"github.com/pulumi/pulumi/pkg/util/logging"
+	"github.com/pulumi/pulumi/pkg/util/result"
+	"github.com/pulumi/pulumi/pkg/workspace"
+	uuid "github.com/satori/go.uuid"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+type lockContent struct {
+	Pid       int       `json:"pid"`
+	Username  string    `json:"username"`
+	Hostname  string    `json:"hostname"`
+	Timestamp time.Time `json:"timestamp"`
+}
+
+func newLockContent() (*lockContent, error) {
+	u, err := user.Current()
+	if err != nil {
+		return nil, err
+	}
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+	return &lockContent{
+		Pid:       os.Getpid(),
+		Username:  u.Username,
+		Hostname:  hostname,
+		Timestamp: time.Now(),
+	}, nil
+}
+
+func (l *lockContent) String() string {
+	return fmt.Sprintf("%v@%v (pid %v) at %v", l.Username, l.Hostname, l.Pid, l.Timestamp.Format(time.RFC3339))
+}
+
+type lockableBackend struct {
+	lb     *localBackend
+	lockID string
+}
+
+func NewLockableBackend(lb *localBackend) Backend {
+	return &lockableBackend{lb: lb, lockID: uuid.NewV4().String()}
+}
+
+func (b *lockableBackend) CreateStack(ctx context.Context, stackRef backend.StackReference,
+	opts interface{}) (backend.Stack, error) {
+	err := b.Lock(stackRef)
+	if err != nil {
+		return nil, err
+	}
+	defer b.Unlock(stackRef)
+	return b.lb.CreateStack(ctx, stackRef, opts)
```

Comment:
so, the way i was envisioning this wsa that there would be a private helper that was basically given the context and stackRef.  It would be passed a function to execute on the inner `b.lb` with the confines of the lock.  That way we wouldn't have to repeat all the code about the locking/unlocking.

Reply:
> Note: i would be happy to complete this PR for you if you want. I can make the small nitty changes and carry it through. Let me know what you'd prefer. If you want to see it to the end, that's totally great as well!

Let it rip!

Reply:
I think I understand what you're saying but I'm not sure how that should best be implemented. The functions have a variety of parameter and response types. 

```
	Refresh(ctx context.Context, stackRef StackReference, op UpdateOperation) (engine.ResourceChanges, result.Result)
	RemoveStack(ctx context.Context, stackRef StackReference, force bool) (bool, error)
	CreateStack(ctx context.Context, stackRef StackReference, opts interface{}) (Stack, error)
```